### PR TITLE
Squad Robo Geniuses, Worksheet feature

### PR DESCRIPTION
**Pull request by Dim314159**
# Description and formatting
We created worksheet generator located features/worksheet

We used tool ID "3".
from api/tools_config.json:
```
    "3": {
        "path": "features.worksheet.core",
        "metadata_file": "metadata.json"
    }
```

Here is the configuration of our input in the features/worksheet/metadata.json:
```
{
    "inputs": [
        {
            "label": "Topic or text",
            "name": "topic",
            "type": "text"
        },
        {
            "label": "Grade level",
            "name": "grade_level",
            "type": "text"
        },
        {
            "label": "Number of Worksheets",
            "name": "num_worksheets",
            "type": "number"
        },
        {
            "label": "Number of Fill in the Blank Questions",
            "name": "num_fill_in_blank",
            "type": "number"
        },
        {
            "label": "Number of Multiple Choice Questions",
            "name": "num_multiple_choice",
            "type": "number"
        },
        {
            "label": "Number of Open-Ended Questions",
            "name": "num_open_ended",
            "type": "number"
        },
        {
            "label": "Number of True or False Questions",
            "name": "num_true_false",
            "type": "number"
        }
    ]
}
```

So our request body may look like:
```
{
    "user": {
        "id": "string",
        "fullName": "string",
        "email": "string"
    },
    "type": "tool",
    "tool_data": {
        "tool_id": 3,
        "inputs": [
            {
                "name": "topic",
                "value": "Visual Arts"
            },
            {
                "name": "grade_level",
                "value": "Undergraduate University (sophomore)"
            },
            {
                "name": "num_worksheets",
                "value": 3
            },
            {
                "name": "num_fill_in_blank",
                "value": 2
            },
            {
                "name": "num_multiple_choice",
                "value": 3
            },
            {
                "name": "num_open_ended",
                "value": 1
            },
            {
                "name": "num_true_false",
                "value": 4
            }
        ]
    }
}
```

'num_worksheets' variable indicate how many worksheets will be created.
The other variables indicate how many fill-in-the-blank, multiple-choice, open-ended and true or false questions will be created in each worksheet.

The result will be the list of dictionaries, where each dictionary represents one worksheet.
Structure of the worksheet dictionary:
- **'description'**: string with description of the topic.
- **'fill_in_blank'**:  list of dictionaries, where each dictionary represents one fill-in-blank question.
   - each fill-in-blank question dictionary has a structure:
      - **'question'**:  question string with removed key word or concept and substituted by "___".
      - **'answer'**: the correct answer (key word or concept removed).
- **'multiple_choice'**: list of dictionaries, where each dictionary represents one multiple-choice question.
   - each multiple-choice question dictionary has a structure:
      - **'question'** : question string.
      - **'choices'**: list of key-value pairs where key is letter and value possible answer.
         - each choice dictionary has a structure:
            - **'key'**: letter (A, B, C, D).
            - **'value'**: possible answer.
      - **'answer'**: the correct choice (A, B, C, D)
- **'open_ended'**: list of dictionaries, where each dictionary represents one open-ended question.
   - each open-ended question dictionary has a structure:
      - **'question'**: question string.
      - **'answer'**: answer string.
- **'true_false'**: list of dictionaries, where each dictionary represents one true or false question.
   - each true or false question dictionary has a structure:
      - **'question'**: question string.
      - **'answer'**: answer string (one word: 'True' or 'False')

# Code structure

in features/worksheet/tools.py we created:
class WorksheetBuilder with method:
- 'create_worksheets', this method generates desired number of worksheets. We used limit:
```
        if num_worksheets > 10:
            return {"message": "error", "data": "Number of Worksheets cannot exceed 10"} 
```
this limit is just for testing purposes and can be eliminated (or changed to different number) for the final version.

For generating questions we created classes:
class QuestionBase with methods:
- 'create_questions' tries to generate indicated amount of questions. 
- 'validate_response' makes sure the generated question obeys format requirements. 
- 'not_unique' makes sure each new question (within one section of the worksheet) is unique. We used embedding model to calculate similarities between new and already created questions.

For each section of the worksheet ('description', 'fill_in_blank', 'multiple_choice', 'open_ended', 'true_false') we created subclass:
class Summary(QuestionBase)
class FillInTheBlankQuestion(QuestionBase)
class MultipleChoiceQuestion(QuestionBase)
class OpenEndedQuestion(QuestionBase)
class TrueFalseQuestion(QuestionBase)
Each subclass has its own 'validate_response' method, since formats of different questions are different.

For the response formatting we used format instructions.
```
prompt = PromptTemplate(
            template=self.prompt_template,
            input_variables=["topic", "grade_level"],
            partial_variables={"format_instructions": self.parser.get_format_instructions()}
        )
```
Each section has its own parser. Some parsers are identical:
```
class TrueFalseQuestionFormat(BaseModel):
    question: str = Field(description = "The question text")
    answer: str = Field(description = "The correct answer")
class OpenEndedQuestionFormat(BaseModel):
    question: str = Field(description = "The question text")
    answer: str = Field(description = "The correct answer")
```
But we decided to keep them in case if in the future we have to adjust them separately, due to possible future formatting demands (or whatever changes can be).

# Testing

We tested worksheet generator for different topics and different grade levels.

Topics: Math, Computer Science, Biology, Philosophy, Chemistry, Literature, Visual Arts, History.
Grade levels: Middle School, High School, Undergraduate University (one of: freshman, sophomore, junior, senior), Graduate School (one of: masters, doctoral), Postdoctoral.

Generator worked well with no significant errors.

# Possible improvements.
for the questions generating we used
"model": VertexAI(model="gemini-1.0-pro", temperature = 0.4)
Sometimes it produces inaccurate questions. For example:
{'question': 'In Shakespeare's play "Romeo and Juliet," Juliet is 13 years old.', 'answer': 'False'}
The correct answer is True. This can be improved by using better model. Or even more complicated way: correctness check through internet search.

### CODE REVIEW

File: app/api/tool_utilities.py [hunk 1]
```diff
@@ -14,18 +14,21 @@ def load_config():
...
         return json.load(f)
 
-tools_config = load_config()
+#tools_config = load_config()
 
 def get_executor_by_name(module_path):
     try:
         module = __import__(module_path, fromlist=['executor'])
-        return getattr(module, 'executor')
+        executor = getattr(module, 'executor')
```
Comment (mikhailocampo) [inline app/api/tool_utilities.py:22]: No need to store as a variable, we can just return

File: app/api/tool_utilities.py [hunk 2]
```diff
@@ -14,18 +14,21 @@ def load_config():
...
+        return executor
     except Exception as e:
         logger.error(f"Failed to import executor from {module_path}: {str(e)}")
         raise ImportError(f"Failed to import module from {module_path}: {str(e)}")
 
 def load_tool_metadata(tool_id):
     logger.debug(f"Loading tool metadata for tool_id: {tool_id}")
+    
+    tools_config = load_config()
```
Comment (mikhailocampo) [inline app/api/tool_utilities.py:31]: Keep tools_config in DRY principle (dont repeat yourself)

File: app/features/quizzify/core.py
Comment (mikhailocampo) [inline app/features/quizzify/core.py:1]: This one will overwrite and remove the quizzify feature!

File: app/features/worksheet/core.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+logger = setup_logger()
+
+def executor(topic: str, grade_level: str, num_worksheets: int, num_fill_in_blank: int, num_multiple_choice: int, num_open_ended: int, num_true_false: int, verbose=False):
+    try:
+        
+        # Create and return the worksheets
+        output = WorksheetBuilder(topic = topic, grade_level = grade_level, verbose=verbose).create_worksheets(num_worksheets = num_worksheets, num_fill_in_blank = num_fill_in_blank, num_multiple_choice = num_multiple_choice, num_open_ended = num_open_ended, num_true_false = num_true_false)
+        
+        ## DEBUG print (remove later)
```
Comment (mikhailocampo) [inline app/features/worksheet/core.py:14]: Debug prints, remember to remove

File: app/features/worksheet/tools.py [hunk 4]
```diff
@@ -0,0 +1,311 @@
+from typing import List, Tuple, Dict, Any
+
+import os
+
+from langchain_google_vertexai import VertexAI
```
Comment (mikhailocampo) [inline app/features/worksheet/tools.py:5]: Should use the class, `from langchain_google_genai import GoogleGenerativeAI` 

Use class, can utilize the API key from aistudio Google

File: app/features/worksheet/tools.py [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+        return generated_questions
+
+# Multiple Choice Question subclass
+class MultipleChoiceQuestion(QuestionBase):
+    def set_names(self):
+        self.section_name = 'Miltiple-Choice question'
+        self.main_field = 'question'
+
+    def validate_response(self) -> bool:
+        try:
```
Comment (mikhailocampo) [inline app/features/worksheet/tools.py:113]: Let's refactor this

Comment (mikhailocampo) [reply app/features/worksheet/tools.py:113]: Research on Pydantic on validations

Comment (mikhailocampo) [reply app/features/worksheet/tools.py:113]: https://refactoring.guru/design-patterns/factory-method

Either pydantic or since your use OOP, try using a Factory Design Pattern

File: app/features/worksheet/tools.py [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+            worksheet["true_false"] = generated_true_false
+
+            generated_worksheets.append(worksheet)
+        # Return the list of worksheets
+        return generated_worksheets
+
+
+class QuestionChoiceFormat(BaseModel):
+    key: str = Field(description="A unique identifier for the choice using letters A, B, C, D, etc.")
+    value: str = Field(description="The text content of the choice")
```
Comment (mikhailocampo) [inline app/features/worksheet/tools.py:292]: @model_validator(args)
def check_value(cls, v): 
# Validate the input fields

Example Pydantic built-in validation for multiple fields

**Review by mikhailocampo — CHANGES_REQUESTED** [review]
Looks good so far, please review changes to be made

File: app/api/tool_utilities.py (hunk 1 above)
Comment (Dim314159) [reply app/api/tool_utilities.py]: Fixed it.

File: app/api/tool_utilities.py (hunk 2 above)
Comment (Dim314159) [reply app/api/tool_utilities.py]: Fixed it.

File: app/features/quizzify/core.py
Comment (Dim314159) [reply app/features/quizzify/core.py:1]: Fixed it.

File: app/features/worksheet/core.py (hunk 3 above)
Comment (Dim314159) [reply app/features/worksheet/core.py:14]: Fixed it.

File: app/features/worksheet/tools.py (hunk 4 above)
Comment (Dim314159) [reply app/features/worksheet/tools.py]: Fixed it.

File: app/features/worksheet/tools.py (hunk 5 above)
Comment (Dim314159) [reply app/features/worksheet/tools.py]: Fixed it. Used pydantic ValidationError.

File: .gitignore [hunk 7]
```diff
@@ -161,4 +161,6 @@ cython_debug/
 
 **/env/
 **/key.txt
-**/local-auth.json
\ No newline at end of file
+**/local-auth.json
+tempo/
```
Comment (AaronSosaRamos) [inline .gitignore:165]: The `tempo` folder and the `works.ipynb` file are not a part of the main project. Please, replace this.

File: .gitattributes [hunk 8]
```diff
@@ -0,0 +1,11 @@
+# Set default behavior to automatically normalize line endings
```
Comment (AaronSosaRamos) [inline .gitattributes:1]: What's the purpose of this change in `.gitattributes`?

File: app/api/tool_utilities.py [hunk 9]
```diff
@@ -125,11 +126,12 @@ def finalize_inputs(input_data, validate_data: List[Dict[str, str]]) -> Dict[str
 
 def execute_tool(tool_id, request_inputs_dict):
     try:
+        tools_config = load_config()
```
Comment (AaronSosaRamos) [inline app/api/tool_utilities.py:129]: Rename this variable as it is very similar to the other variable: `tool_config`. This is good for having an intuitive understanding of the main flow of the code.

File: app/api/tools_config.json [hunk 10]
```diff
@@ -6,5 +6,13 @@
     "1": {
         "path": "features.dynamo.core",
         "metadata_file": "metadata.json"
+    },
+    "2": {
+        "path": "features.code_check.core",
```
Comment (AaronSosaRamos) [inline app/api/tools_config.json:11]: This feature is not required as it is not a part of the **Worksheet Generator** feature.

File: app/features/worksheet/core.py [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+from features.worksheet.tools import WorksheetBuilder
+from api.error_utilities import LoaderError, ToolExecutorError
+
+logger = setup_logger()
+
+def executor(topic: str, grade_level: str, num_worksheets: int, num_fill_in_blank: int, num_multiple_choice: int, num_open_ended: int, num_true_false: int, verbose=False):
+    try:
+        
+        # Create and return the worksheets
+        output = WorksheetBuilder(topic = topic, grade_level = grade_level, verbose=verbose).create_worksheets(num_worksheets = num_worksheets, num_fill_in_blank = num_fill_in_blank, num_multiple_choice = num_multiple_choice, num_open_ended = num_open_ended, num_true_false = num_true_false)
```
Comment (AaronSosaRamos) [inline app/features/worksheet/core.py:12]: Try to use multiple lines for a better readability.

File: app/features/worksheet/core.py [hunk 12]
```diff
@@ -0,0 +1,24 @@
+from services.tool_registry import ToolFile
+from services.logger import setup_logger
+from features.worksheet.tools import WorksheetBuilder
+from api.error_utilities import LoaderError, ToolExecutorError
+
+logger = setup_logger()
+
+def executor(topic: str, grade_level: str, num_worksheets: int, num_fill_in_blank: int, num_multiple_choice: int, num_open_ended: int, num_true_false: int, verbose=False):
```
Comment (AaronSosaRamos) [inline app/features/worksheet/core.py:8]: You can decouple this args. using a Pydantic Schema.

File: app/features/worksheet/prompts/worksheet_prompt_multiple_choice_old.txt [hunk 13]
```diff
@@ -0,0 +1,8 @@
+You are a subject matter expert on the topic: {topic}.
```
Comment (AaronSosaRamos) [inline app/features/worksheet/prompts/worksheet_prompt_multiple_choice_old.txt:1]: If this is an old prompt, you can remove it or consider it as **deprecated**. Legacy components are not considered a part of the main workflow unless you're transitioning from different versions of your code.

File: app/features/worksheet/tools.py [hunk 14]
```diff
@@ -0,0 +1,278 @@
...
+
+    # Combine the script directory with the relative file path
+    absolute_file_path = os.path.join(script_dir, file_path)
+    
+    with open(absolute_file_path, 'r') as file:
+        return file.read()
+
+# base class
+class QuestionBase:
+    def __init__(self, model, embedding_model, topic, grade_level, prompt_template, parser, verbose):
```
Comment (AaronSosaRamos) [inline app/features/worksheet/tools.py:29]: You can try decoupling these args. using a Pydantic Schema.

File: app/features/worksheet/tools.py [hunk 15]
```diff
@@ -0,0 +1,278 @@
...
+        return False
+    
+    def create_questions(self, num_questions = 0) -> List:
+        attempts = 0
+        max_attempts = num_questions * 5  # Allow for more attempts to generate questions
+        self.question_bank = []
+        self.existing_questions = ''
+        generated_questions = []
+
+        while len(generated_questions) < num_questions and attempts < max_attempts:
```
Comment (AaronSosaRamos) [inline app/features/worksheet/tools.py:96]: Ensure that this loop doesn't generate too many tokens as it can be considered as costful.

File: app/features/worksheet/tools.py [hunk 16]
```diff
@@ -0,0 +1,278 @@
...
+                    logger.warning(f"Invalid {self.section_name} response format. Attempt {attempts} of {max_attempts}")
+        
+        # Log if fewer questions are generated
+        if len(generated_questions) < num_questions:
+            logger.warning(f"Only generated {len(generated_questions)} out of {num_questions} requested {self.section_name}(s)")
+
+        return generated_questions
+
+# Multiple Choice Question subclass
+class MultipleChoiceQuestion(QuestionBase):
```
Comment (AaronSosaRamos) [inline app/features/worksheet/tools.py:126]: Good practice to use OOP for reusability

File: app/features/worksheet/tools.py [hunk 17]
```diff
@@ -0,0 +1,278 @@
...
+        self.question_bank = []
+        self.existing_questions = ''
+        generated_questions = []
+
+        while len(generated_questions) < num_questions and attempts < max_attempts:
+            # Move to the next attempt regardless of success to ensure progress
+            attempts += 1
+            self.response = self.chain.invoke({"topic": self.topic, "existing_questions": self.existing_questions, "grade_level": self.grade_level})
+
+            if self.verbose:
```
Comment (AaronSosaRamos) [inline app/features/worksheet/tools.py:101]: Verbose is not logged. You can use print statements instead.

File: app/features/worksheet/tools.py [hunk 18]
```diff
@@ -0,0 +1,278 @@
...
+        self.section_name = 'True or False question'
+        self.main_field = 'question'
+    
+    def sub_validate(self):
+        TrueFalseQuestionFormat(**self.response)
+        return True
+
+# Create worksheets
+class WorksheetBuilder:
+    def __init__(self, topic, grade_level, model=None, embedding_model = None, prompt_summary=None, parser_summary = None, prompt_fill_in_blank=None, parser_fill_in_blank=None, prompt_multiple_choice=None, parser_multiple_choice=None, prompt_open_ended = None, parser_open_ended = None, prompt_true_false = None, parser_true_false = None, verbose=False):
```
Comment (AaronSosaRamos) [inline app/features/worksheet/tools.py:182]: Use a Pydantic Schema for decoupling args. and ensuring a better reusability and readability.

File: app/features/worksheet/tools.py [hunk 19]
```diff
@@ -0,0 +1,278 @@
...
+        if topic is None or grade_level is None:
+            raise ValueError("Topic and Grade level must be provided")
+        
+        default_config = {
+            "model": GoogleGenerativeAI(model="gemini-1.0-pro", temperature=0.5),
+            "embedding_model": SentenceTransformer('all-MiniLM-L6-v2'),
+            "prompt_summary": read_text_file("prompts/worksheet_prompt_summary.txt"),
```
Comment (AaronSosaRamos) [inline app/features/worksheet/tools.py:189]: Divide the `prompt_` attributes for keeping the Single Responsability Principle

File: app/features/worksheet/tools.py [hunk 20]
```diff
@@ -0,0 +1,278 @@
...
+            "parser_open_ended": JsonOutputParser(pydantic_object = OpenEndedQuestionFormat),
+            "prompt_true_false": read_text_file("prompts/worksheet_prompt_true_false.txt"),
+            "parser_true_false": JsonOutputParser(pydantic_object = TrueFalseQuestionFormat)
+        }
+        
+        self.model = model or default_config["model"]
+        self.embedding_model = embedding_model or default_config["embedding_model"]
+        self.prompt_summary = prompt_summary or default_config["prompt_summary"]
+        self.parser_summary = parser_summary or default_config["parser_summary"]
+        self.prompt_fill_in_blank = prompt_fill_in_blank or default_config["prompt_fill_in_blank"]
```
Comment (AaronSosaRamos) [inline app/features/worksheet/tools.py:205]: Use another method for reusing the components and having a better scalability. Don't generate variables for each value, reuse the previous ones.

File: app/features/worksheet/tools.py [hunk 21]
```diff
@@ -0,0 +1,278 @@
...
+        self.grade_level = grade_level
+        self.verbose = verbose
+    
+    def create_worksheets(self, num_worksheets: int = 1, num_fill_in_blank: int = 1, num_multiple_choice: int = 1, num_open_ended: int = 1, num_true_false: int = 1) -> List[str]:
+        ### Limit just for testing
+        if num_worksheets > 10:
+            return {"message": "error", "data": "Number of Worksheets cannot exceed 10"}
+        
+        summary_obj = Summary(self.model, self.embedding_model, self.topic, self.grade_level, self.prompt_summary, self.parser_summary, self.verbose)
+        fill_in_the_blank_obj = FillInTheBlankQuestion(self.model, self.embedding_model, self.topic, self.grade_level, self.prompt_fill_in_blank, self.parser_fill_in_blank, self.verbose)
```
Comment (AaronSosaRamos) [inline app/features/worksheet/tools.py:224]: Use a Pydantic Schema for reuse the args as well as separating this in another function for not repeating the same parts of the code.

File: app/features/worksheet/tools.py [hunk 22]
```diff
@@ -0,0 +1,278 @@
...
+        generated_summary = summary_obj.create_questions(num_questions = 1)
+        if generated_summary:
+            generated_summary = generated_summary[0]
+        else:
+            generated_summary = {'description': ''}
+        for i in range(num_worksheets):
+            worksheet = {}
+            worksheet.update(generated_summary)
+            if self.verbose: logger.info(f"Creating {num_fill_in_blank} Fill-in-the-blank questions")
+            generated_fill_in_blank = fill_in_the_blank_obj.create_questions(num_questions = num_fill_in_blank)
```
Comment (AaronSosaRamos) [inline app/features/worksheet/tools.py:239]: The same for this.

File: app/features/worksheet/tools.py [hunk 23]
```diff
@@ -0,0 +1,278 @@
...
+            generated_worksheets.append(worksheet)
+        # Return the list of worksheets
+        if self.verbose: logger.info(f"Created {len(generated_worksheets)} Worksheets")
+        return generated_worksheets
+
+
+class QuestionChoiceFormat(BaseModel):
+    key: str = Field(description="A unique identifier for the choice using letters A, B, C, D, etc.")
+    value: str = Field(description="The text content of the choice")
+class MultipleChoiceQuestionFormat(BaseModel):
```
Comment (AaronSosaRamos) [inline app/features/worksheet/tools.py:260]: For more info. about input validation, check this doc: https://docs.pydantic.dev/latest/concepts/validators/#ordering-of-validators-within-annotated

File: app/features/worksheet/tools.py (hunk 20 above)
Comment (AaronSosaRamos) [reply app/features/worksheet/tools.py:205]: DRY

File: app/features/worksheet/tools.py (hunk 22 above)
Comment (AaronSosaRamos) [reply app/features/worksheet/tools.py:239]: DRY

**Review by AaronSosaRamos — CHANGES_REQUESTED** [review]
Please, solve the comments and let us know.

**Comment by AaronSosaRamos** [comment]
Closed due inactivity
